dlrover/python/master/elastic_training/net_topology.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2024 The DLRover Authors. All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from dataclasses import dataclass
+from typing import Dict, List, Tuple
+
+
+@dataclass
+class NodeTopologyMeta(object):
+    node_rank: int = 0
+    process_num: int = 0
+    node_ip: str = ""
+    asw: str = ""
+    psw: str = ""
+
+
+class TopologyQuerier(metaclass=ABCMeta):
+    @abstractmethod
+    def query(self, node_ip) -> Tuple[str, str]:
+        """Query the asw and psw id of a node by the IP."""
+        pass
+
+
+class TopologySorter(metaclass=ABCMeta):
+    @abstractmethod
+    def sort(
+        self, nodes: Dict[int, NodeTopologyMeta]
+    ) -> Dict[int, NodeTopologyMeta]:
+        """Query the asw and psw id of a node by the IP."""
+        pass
+
+
+class DefaultTopologyQuerier(TopologyQuerier):
+    def query(self, node_ip) -> Tuple[str, str]:
+        return "", ""
+
+
+class DpTopologySorter(TopologySorter):
+    """
+    The sorter places the nodes under an asw (access switch) together in
+    the list of nodes. In allreduce communication, the communication packets
+    between nodes with continuous ranks under an asw will not pass the psw.
+
+    """
+
+    def sort(
```

Comment:
I'm not sure. Is the following implementation adapt most network arch on kubernetes?